### PR TITLE
gnrc_ndp_internal: send RAs even when there are no PIOs

### PR DESCRIPTION
**Pull request by authmillenon**
RAs aren't currently send if there is no global unicast address set. This is because the function that builds the PIOs both assumes NULL for the error case and an empty PIO list, resulting in the RA send function to drop the packet altogether.

### CODE REVIEW

File: sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c [hunk 1]
```diff
@@ -406,24 +404,24 @@ static gnrc_pktsnip_t *_add_pios(gnrc_ipv6_netif_t *ipv6_iface, gnrc_pktsnip_t *
             continue;
         }
 
-        if (_pio_from_iface_addr(&tmp, &ipv6_iface->addrs[i], pkt)) {
-            if (tmp != NULL) {
-                pkt = tmp;
+        if (_pio_from_iface_addr(res, &ipv6_iface->addrs[i], pkt)) {
+            if (*res != NULL) {
+                pkt = *res;
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c]: this line is unnecessary

Comment (authmillenon) [inline sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c]: no, otherwise pkt won't change.

Comment (authmillenon) [inline sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c]: (resulting in memory leaks in https://github.com/RIOT-OS/RIOT/pull/4174/files#diff-eb98861cc04f206d3376ec33720dfcc7R413)

Comment (authmillenon) [inline sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c]: And only the last PIO being added to the packet.

**Comment by cgundogan** [comment]
ACK

**Comment by OlegHahm** [comment]
Wait, doesn't it make sense to not send RAs if you have no routable address?

**Comment by authmillenon** [comment]
Why? Only because you have no routable address on the advertising interface doesn't mean you have none on another.

**Comment by authmillenon** [comment]
It doesn't make sense to not advertise however, only because there is no prefix set (yet).

**Comment by OlegHahm** [comment]
So, one only send if you have a routable prefix on any interface with this PR? In this case I'm fine.

**Comment by authmillenon** [comment]
One sends RAs if the router flag and the rtr_adv flag is set. As specified by the RFC. Checking if any other interface has routable prefixes set would mean unnecessary overhead for something that should be in discretion of the administrator.

**Comment by OlegHahm** [comment]
A 6LR won't have no administrator and trading some increased runtime overhead against not sending pointless packets is worth it for energy constrained devices. I haven't checked if the RFC states that one MUST send RAs even if there is no routable prefix configured, but even if it does I would consider energy efficiency more important than dogmatic RFC compliance.

**Comment by authmillenon** [comment]
Than you can introduce that case. The current behavior is however buggy/not as it is intended.

**Comment by authmillenon** [comment]
> A 6LR won't have no administrator

Huh? So prefixes just magically appear in the 6LoWPAN?

**Comment by OlegHahm** [comment]
> Huh? So prefixes just magically appear in the 6LoWPAN?

No, but they are disseminated through a 6LBR.

**Comment by OlegHahm** [comment]
IMO the current situation is better, because no RAs won't be sent before a routable prefix is disseminated which saves some energy. This is IMO a good behaviour for all cases except the one where you have multiple interfaces. In at least 99% of all RIOT scenarios this is currently only the case for 6LBRs and there it is of limited use to send RAs if no prefix has been assigned to the 6lo interface or am I mistaken?

**Comment by authmillenon** [comment]
I'm sorry, but arguing that a bug should be kept in the code because it saves energy is poor design. If you want to have this behavior please provide some better readable code than this. Furthermore, a 6LR doesn't start advertising [until itself received an RA from an upstream router](https://tools.ietf.org/html/rfc6775#section-6.2) (with the 6LBR being the most upstream), and if this does not carry any PIOs than the 6LoWPAN is ill-configured anyway.

**Comment by OlegHahm** [comment]
If it saves energy, it is not a bug. I disagree.

**Comment by authmillenon** [comment]
It is a bug, or faulty design, if you want to be precise. We can e.g. do it RFC comforming AND the way you want it by unsetting the RTR_ADV flag, as long as there is no GUA set. Then the RA send function wouldn't even be called and not just discard an already partly executed process due to a faulty return value.

**Comment by OlegHahm** [comment]
Sounds like a compromise. What happens if the a prefix (i.e. the last routable prefix) is removed and RAs are being already sent. Are there any timers that might still cause the router to send RAs?

**Comment by authmillenon** [comment]
In a LoWPAN yes. An RS from a 6LN solicitates RAs from the routers, but those are also necessary for address registration.

**Comment by authmillenon** [comment]
Ah yes… that's another thing. If there are no routers AR wouldn't work. And that's independent of the router having a GUA (only link-local addresses based on the EUI-64 are considered to be unique and need not be registered). So things like 6LoBAC or 6Locc11xx only would work if you bootstrap a network from a fully configured border router up … which is kind of lame, because without this RA restriction it would just work.

**Comment by authmillenon** [comment]
with "If there are no routers" I meant "no nodes that advertise themselves as routers"

**Comment by OlegHahm** [comment]
I don't get it. What kind of AR should a 6lowpan node perform without a prefix?

**Comment by authmillenon** [comment]
For link-local addresses based on 8- or 16-bit short addresses.

**Comment by OlegHahm** [comment]
IMO it makes sense for a node not to send any AROs until it knows that there is router. (If the router does not have a routable prefix, it's not a router.) Of course, you establish link local communication between these nodes without a routable prefix, but a) is "link local" not a very meaningful term in wireless networks and b) I cannot really see a use case for this.

**Comment by authmillenon** [comment]
on a) true, but that makes it in my eyes to an "arbitrary" address ready to use in a wireless network, and this thinking seems to align with the design of 6Lo-ND and its handling of link-layer addresses, IMO.
on b) simple "slice-of-the-big-cake" test scenarios would be impossible if you always have to set up a 6LBR first.

**Comment by OlegHahm** [comment]
Discussed this issue just with @cabo and he pointed me to [RFC6775, section 3.2](https://tools.ietf.org/html/rfc6775#section-3.2) which basically says that a) an 6LBR should generate a ULA that can be used in case that connectivity to the uplink is temporarily or permanently broken (should be included in RAs as well) and b) a 6LR might turn into a 6LBR if there is no 6LBR present in the network by using a ULA prefix.

**Comment by authmillenon** [comment]
So what does this mean for this PR?

**Comment by OlegHahm** [comment]
I still think that in the single-interface case. it doesn't make sense to send RAs at all, when you don't have a prefix to advertise. However, I agree it might make sense in a case where you have more than one interface. Additionally, we might want to a) set a timer to generate a ULA after some time if no routable prefix is configured on a router and b) make a 6LR act as a 6LBR in case no 6LBRs shows up.

**Comment by authmillenon** [comment]
Then we might want to go with what I proposed in https://github.com/RIOT-OS/RIOT/pull/4174#issuecomment-151621768 merge this PR so the unnecessary function calling is stopped.

**Comment by OlegHahm** [comment]
sounds good

**Comment by authmillenon** [comment]
See #4252.

**Comment by authmillenon** [comment]
Sorry… accidentally hit the wrong button >.<

**Comment by authmillenon** [comment]
@OlegHahm just to be clear: the ACK by @cgundogan upholds?

**Comment by cgundogan** [comment]
yes it does. ACK

**Comment by authmillenon** [comment]
Travis is happy too

**Pull request merged**